component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,352 @@
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	delete(p.openStores, name)
+}
+
+type store struct {
+	name            string
+	underlyingStore newstorage.Store
+	currentBatch    []newstorage.Operation
+	batchSizeLimit  int
+	close           closer
+	*sync.RWMutex
+}
+
+func (s *store) Put(key string, value []byte, tags ...newstorage.Tag) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	if value == nil {
+		return errors.New("value cannot be nil")
+	}
+
+	s.Lock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: value,
+		Tags:  tags,
+	})
+
+	currentBatchLength := len(s.currentBatch)
+
+	s.Unlock()
+
+	if currentBatchLength >= s.batchSizeLimit {
+		err := s.Flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+// TODO (#2485): Check data in current batch before resorting to a flush.
+func (s *store) Get(key string) ([]byte, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	value, err := s.underlyingStore.Get(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get value from underlying store: %w", err)
+	}
+
+	return value, nil
+}
+
+func (s *store) GetTags(key string) ([]newstorage.Tag, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	tags, err := s.underlyingStore.GetTags(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tags from underlying store: %w", err)
+	}
+
+	return tags, nil
+}
+
+func (s *store) GetBulk(keys ...string) ([][]byte, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	values, err := s.underlyingStore.GetBulk(keys...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return values, nil
+}
+
+func (s *store) Query(expression string, options ...newstorage.QueryOption) (newstorage.Iterator, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	iterator, err := s.underlyingStore.Query(expression, options...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return iterator, nil
+}
+
+func (s *store) Delete(key string) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	s.Lock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: nil,
+	})
+
+	currentBatchLength := len(s.currentBatch)
+
+	s.Unlock()
+
+	if currentBatchLength >= s.batchSizeLimit {
+		err := s.Flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+func (s *store) Batch(operations []newstorage.Operation) error {
+	for _, operation := range operations {
+		if operation.Key == "" {
+			return errors.New("an operation's key was empty")
+		}
+	}
+
+	for _, operation := range operations {
+		s.Lock()
+
+		s.currentBatch = append(s.currentBatch, newstorage.Operation{
```

Comment:
@Moopli Thanks for catching this - I don't know why I copied the operation object manually like this. I should just be appending `operation` here. I will fix it now